minerl/herobraine/env_spec.py

```diff
@@ -1,103 +1,118 @@
+        self._action_space = self.create_action_space()
+
+    @property
+    def observation_space(self):
+        return self._observation_space
+
+    @property
+    def action_space(self):
+        return self._action_space
+
+    def to_string(self):
+        return self.name
+
+    @abstractmethod
+    def is_from_folder(self, folder: str) -> bool:
+        raise NotImplementedError('subclasses must override is_from_folder()!')
+
+    @abstractmethod
+    def create_mission_handlers(self):
+        raise NotImplementedError('subclasses must override create_mission_handlers()!')
+
+    @abstractmethod
+    def create_observables(self):
+        raise NotImplementedError('subclasses must override create_observables()!')
+
+    @abstractmethod
+    def create_actionables(self):
+        raise NotImplementedError('subclasses must override create_actionables()!')
+
+    @abstractmethod
+    def determine_success_from_rewards(self, rewards: list) -> bool:
+        raise NotImplementedError('subclasses must override determine_success_from_rewards()')
+
+    def auto_blacklist(self, npz_data: dict) -> Optional[str]:
+        """Return a non-empty str if `publish.py` should blacklist a demonstration.
+
+        We can't catch all cases of bad demonstrations automatically, but overriding this
+        method can allow for some quick, automatic filtering.
+
+        Args:
+            npz_data: A dict of numpy values about to be saved as "rendered.npz".
+
+        Returns:
+            Either None, or a nonempty str describing why this demonstration should be
+            blacklisted.
+        """
+        return None
+
+    def create_observation_space(self):
+        # Todo: handle nested dict space.
+        return spaces.Dict({
+            o.to_string(): o.space for o in self.observables
+        })
+
+    def create_action_space(self):
+        return spaces.Dict({
+            a.to_string(): a.space for a in self.actionables
+        })
+
+    @abstractmethod
+    def get_docstring(self):
+        return NotImplemented
+
+    def register(self):
+        reg_spec = dict(
+            id=self.name,
+            entry_point=EnvSpec.ENTRYPOINT,
+            kwargs={
+                'observation_space': self.observation_space,
+                'action_space': self.action_space,
+                'docstr': self.get_docstring(),
+                'xml': os.path.join(MISSIONS_DIR, self.xml),
+                'env_spec': self,
+            },
+            max_episode_steps=self.max_episode_steps,
+        )
+        if self.reward_threshold:
+            reg_spec.update(dict(reward_threshold=self.reward_threshold))
+
+        gym.register(**reg_spec)
+
+    def __repr__(self):
+        """
+        Prints the class, name, observation space, and action space of the handler.
+        """
+        return '{}-{}-spaces({},{})'.format(self.__class__.__name__, self.name, self.observation_space,
+                                            self.action_space)
```

Comment:
Why is git generating a massive diff here?

Reply:
I used `dos2unix` to convert this file from dos to unix format (all the other files are in dos format). Basically, I converted the end of each line to the standard `\n` instead of whatever dos uses, which was making `git` commands print angry red carriage characters